### DISCUSSION TITLE
Need Help in ellipse coordinates

### DISCUSSION THREAD

**Discussion by Adeeljavedvirk** [openlayers/openlayers#14914] [General]
Hi i want to draw the ellipse shape.  Here is the code snippet which i am using for ellipse shape. 
```js
            drawInteraction = new Draw({
                source: vectorSource,
                type: 'Circle',
                geometryFunction: function (coordinates, geometry) {
                    const center: any = coordinates[0]
                    const last: any = coordinates[1]
                    const dx = center[0] - last[0]
                    const dy = center[1] - last[1]
                    const radius = Math.sqrt(dx * dx + dy * dy)
                    const circle = new Circle(center, radius)
                    const polygon = fromCircle(circle, 64)
                    polygon.scale(dx / radius, dy / radius)
                    if (!geometry) {
                        geometry = polygon
                    } else {
                        geometry.setCoordinates(polygon.getCoordinates())
                    }
                    return geometry
                }
            })
``` 

I can draw the perfect shape with this code but the issue is when i am trying to save this shape i got the 64 coordinates against this code but i want to get just three coordinates because in my old application i can see just three coordinates on save and if i use that three coordinates i can see the triangle shape on map in angular applcation but in old application i can the ellipse shape. I am converting the application from Dojo to angular so it's hard for me to debug how three coordinates are returning from dojo application

**Comment by ahocevar** [comment]
If the 3 points of your triangle are center, minor radius and major radius, I'd configure the Draw interaction with a `Style` with a `geometry` instead of a `geometryFunction`, and calculate the ellipse from these 3 points on the fly. The geometry resulting from drawing would then become a polygon, you could give it an attribute to mark it as ellipse. Same for the layer that displays ellipses that are defined with these 3 points.

**Reply by Adeeljavedvirk** [reply]
If possible can you please share the working or any reference example. Which i can follow

**Comment by ahocevar** [comment]
> If possible can you please share the working or any reference example. Which i can follow

I think you should first provide an example of how an ellipse is defined in your current application, and the visual result thereof. I want to confirm what the 3 points are.

**Reply by Adeeljavedvirk** [reply]
Here is the shapePositionInfo which we use to load the ellipse shape on map.
`'{\\\"shapeType\\\":\\\"Ellipse\\\",\\\"coordinates\\\":[[393.8466343901297,718.2008356335348],[613.1188505388101,718.8885493888301],[393.8466343901297,788.3679293588272]],\\\"style\\\":{\\\"opacity\\\":100,\\\"color\\\":{\\\"value\\\":\\\"Transparent\\\",\\\"id\\\":1,\\\"version\\\":\\\"808080808a811df3\\\",\\\"translations\\\":null,\\\"columns\\\":[{\\\"name\\\":\\\"ColorLabel\\\",\\\"id\\\":\\\"6085\\\",\\\"value\\\":\\\"Transparent\\\"},{\\\"name\\\":\\\"Alpha\\\",\\\"id\\\":\\\"6089\\\",\\\"value\\\":\\\"0.0\\\"},{\\\"name\\\":\\\"Red\\\",\\\"id\\\":\\\"6086\\\",\\\"value\\\":\\\"0\\\"},{\\\"name\\\":\\\"Green\\\",\\\"id\\\":\\\"6087\\\",\\\"value\\\":\\\"0\\\"},{\\\"name\\\":\\\"Blue\\\",\\\"id\\\":\\\"6088\\\",\\\"value\\\":\\\"0\\\"}],\\\"label\\\":\\\"Transparent\\\",\\\"a\\\":0,\\\"r\\\":0,\\\"g\\\":0,\\\"b\\\":0},\\\"stroke\\\":{\\\"value\\\":\\\"Red\\\",\\\"id\\\":2,\\\"version\\\":\\\"808080808a811df4\\\",\\\"translations\\\":null,\\\"columns\\\":[{\\\"name\\\":\\\"ColorLabel\\\",\\\"id\\\":\\\"6085\\\",\\\"value\\\":\\\"Red\\\"},{\\\"name\\\":\\\"Alpha\\\",\\\"id\\\":\\\"6089\\\",\\\"value\\\":\\\"1.0\\\"},{\\\"name\\\":\\\"Red\\\",\\\"id\\\":\\\"6086\\\",\\\"value\\\":\\\"-1\\\"},{\\\"name\\\":\\\"Green\\\",\\\"id\\\":\\\"6087\\\",\\\"value\\\":\\\"0\\\"},{\\\"name\\\":\\\"Blue\\\",\\\"id\\\":\\\"6088\\\",\\\"value\\\":\\\"0\\\"}],\\\"label\\\":\\\"Red\\\",\\\"a\\\":1,\\\"r\\\":255,\\\"g\\\":0,\\\"b\\\":0},\\\"lineWidth\\\":{\\\"id\\\":2,\\\"label\\\":\\\"2px\\\",\\\"value\\\":2},\\\"lineStyleValue\\\":1,\\\"textSizeValue\\\":1,\\\"textValue\\\":12,\\\"textFill\\\":{\\\"value\\\":\\\"Black\\\",\\\"id\\\":13,\\\"version\\\":\\\"808080808a811dff\\\",\\\"translations\\\":null,\\\"columns\\\":[{\\\"name\\\":\\\"ColorLabel\\\",\\\"id\\\":\\\"6085\\\",\\\"value\\\":\\\"Black\\\"},{\\\"name\\\":\\\"Alpha\\\",\\\"id\\\":\\\"6089\\\",\\\"value\\\":\\\"1.0\\\"},{\\\"name\\\":\\\"Red\\\",\\\"id\\\":\\\"6086\\\",\\\"value\\\":\\\"0\\\"},{\\\"name\\\":\\\"Green\\\",\\\"id\\\":\\\"6087\\\",\\\"value\\\":\\\"0\\\"},{\\\"name\\\":\\\"Blue\\\",\\\"id\\\":\\\"6088\\\",\\\"value\\\":\\\"0\\\"}],\\\"label\\\":\\\"Black\\\",\\\"a\\\":1,\\\"r\\\":0,\\\"g\\\":0,\\\"b\\\":0},\\\"stampAuthValue\\\":{\\\"value\\\":\\\"1\\\",\\\"id\\\":1,\\\"version\\\":\\\"808080808b75703c\\\",\\\"translations\\\":null,\\\"columns\\\":[{\\\"name\\\":\\\"KeyValue\\\",\\\"id\\\":\\\"6074\\\",\\\"value\\\":\\\"1\\\"}]},\\\"zIndex\\\":10001},\\\"rotationDegrees\\\":0,\\\"rotationCenter\\\":[0,0],\\\"text\\\":\\\"\\\",\\\"editorText\\\":[]}'`

Here is the visual result of above code on map
![image](https://github.com/openlayers/openlayers/assets/19381997/6d3a0a48-9246-4965-8f39-791be069a452)

**Reply by ahocevar** [reply]
So, your coordinates are
```js
const coordinates = [
   [393.8466343901297, 718.2008356335348],
   [613.1188505388101, 718.8885493888301],
   [393.8466343901297, 788.3679293588272]
]
```
To extract the ellipse parameters, use
```js
const center = coordinates[0];
const dx = coordinates[1][0] - coordinates[0][0];
const dy = coordinates[2][1] - coordinates[0][1];
```
Now assuming your features have a `LineString` type `geometry` with this kind of coordinates, and a `shapeType` property with the value `Ellipse`, you can draw them in OpenLayers as ellipse, using a `Style` with a `geometry`:
```js
const ellipseStyle = new Style({
  geometry(feature) {
    const geometry = feature.getGeometry();
    if (feature.get('shapeType') === 'Ellipse') {
      const coordinates = geometry.getCoordinates();
      // calculate center, dx and dy like shown above
      const circle = new Circle(center, dx)
      const polygon = fromCircle(circle, 64)
      polygon.scale(1, dy / dx);
      return polygon;
    } else {
      return geometry;
    }
  },
  stroke: new Stroke({
    color: 'red',
    width: 2
  });
});
```
And in the `Draw` interaction, to create a `LineString` geometry with 3 points representing an ellipse, modify your `geometryFunction` to something like this:
```js
geometryFunction: function (coordinates, geometry) {
  const center: any = coordinates[0]
  const last: any = coordinates[1]
  const dx = center[0] - last[0]
  const dy = center[1] - last[1]
  const radius = Math.sqrt(dx * dx + dy * dy)
  const lineString = new LineString([center, [center[0] + dx, center[1]], [center[0], center[1] + dy]]]);
  if (!geometry) {
    geometry = lineString
  } else {
    geometry.setCoordinates(lineString.getCoordinates())
  }
  return geometry
}
```

**Reply by Adeeljavedvirk** [reply]
Thanks for your answer. I successfully load the ellipse with ellipseStyle geometry but i am getting the one error in geometryFunction. When i am trying to draw the ellipse i am getting the error "lineString() is not a function". I looked the openlayer documentation i saw the lineString.getCoordinates() function for getting the coordinates but still shape is not properly drawn. 
Here is the code which i am using right now 
```
drawInteraction = new Draw({
                source: vectorSource,
                type: 'Circle',
                geometryFunction: function (coordinates, geometry) {
                    const center: any = coordinates[0]
                    const last: any = coordinates[1]
                    const dx = center[0] - last[0]
                    const dy = center[1] - last[1]
                    const radius = Math.sqrt(dx * dx + dy * dy)
                    const lineString: any = new LineString([center, [center[0] + dx, center[1]], [center[0], center[1] + dy]]);
                    if (!geometry) {
                      geometry = lineString
                    } else {
                      geometry.setCoordinates(lineString.getCoordinates())
                    }
                    return geometry
                  }
```

and here is the result on map with this code.
![image](https://github.com/openlayers/openlayers/assets/19381997/356852fd-75df-4c80-89ef-6e432bf6ba5f)

**Reply by ahocevar** [reply]
You're right, I meant `lineString.getCoordinates()`. I updated my answer above.

The reason why you're seeing a triangle instead of an ellipse is unrelated to that change though. I suspect it is because the contents of your vector source is different from my assumptions, which I clearly stated above already: features have a LineString geometry, and a `shapeType` property. if that's not the case, and all geometries you're drawing are ellipses anyway, you could simplify the `geometry` function in your style:
```js
geometry(feature) {
  const geometry = feature.getGeometry();
  const coordinates = geometry.getCoordinates();
  // calculate center, dx and dy like shown above
  const circle = new Circle(center, dx)
  const polygon = fromCircle(circle, 64)
  polygon.scale(1, dy / dx);
  return polygon;
}
```

**Reply by Adeeljavedvirk** [reply]
I can load the ellipse with above geometry but now i am facing issue in draw if i use this code for draw i got the 64 coordinates.
```
            drawInteraction = new Draw({
                source: vectorSource,
                type: 'Circle',
                geometryFunction: function (coordinates, geometry) {
                    const center: any = coordinates[0]
                    const last: any = coordinates[1]
                    const dx = center[0] - last[0]
                    const dy = center[1] - last[1]
                    const radius = Math.sqrt(dx * dx + dy * dy)
                    const circle = new Circle(center, radius)
                    const polygon = fromCircle(circle, 64)
                    polygon.scale(dx / radius, dy / radius)
                    if (!geometry) {
                        geometry = polygon
                    } else {
                        geometry.setCoordinates(polygon.getCoordinates())
                    }
                    return geometry
                }
            })
```

How i can draw the ellipse and get the three coordinates on draw end ?

**Reply by ahocevar** [reply]
That's the wrong `geometryFunction`. You need the one I provided above.

**Reply by Adeeljavedvirk** [reply]
![image](https://github.com/openlayers/openlayers/assets/19381997/d4a88584-f18a-4a87-a762-4b16690a7ec1)

Red ellipse is showing on the map when i used this geometry . It's working fine as expected.
```
geometry(feature) {
  const geometry = feature.getGeometry();
  const coordinates = geometry.getCoordinates();
  // calculate center, dx and dy like shown above
  const circle = new Circle(center, dx)
  const polygon = fromCircle(circle, 64)
  polygon.scale(1, dy / dx);
  return polygon;
}
```

Blue triangle when i am trying to draw new ellipse with mouse on map i got triangle instead of ellipse with this geometry function. 
```
drawInteraction = new Draw({
                source: vectorSource,
                type: 'Circle',
                geometryFunction: function (coordinates, geometry) {
                    const center: any = coordinates[0]
                    const last: any = coordinates[1]
                    const dx = center[0] - last[0]
                    const dy = center[1] - last[1]
                    const radius = Math.sqrt(dx * dx + dy * dy)
                    const lineString: any = new LineString([center, [center[0] + dx, center[1]], [center[0], center[1] + dy]]);
                    if (!geometry) {
                      geometry = lineString
                    } else {
                      geometry.setCoordinates(lineString.getCoordinates())
                    }
                    return geometry
                  }
```

I need your help how i can draw new ellipse and get just three coordinates on feature save.

**Reply by Adeeljavedvirk** [reply]
const radius = Math.sqrt(dx * dx + dy * dy)

radius variable is not used anywhere. Its expected or need to add somewhere but i am still getting the blue shape with your geometry function

![Screenshot_1](https://github.com/openlayers/openlayers/assets/19381997/d97e9e78-07a2-4d14-9d0d-98d51364645c)

**Reply by ahocevar** [reply]
> I need your help how i can draw new ellipse and get just three coordinates on feature save.

This is exactly what the above function does. Note there are two parts of the equation: the representation of the eclipse in the vector source (which you want to be 3 points), and how it's rendered. For the former, the above `geometryFunction` is needed, For the latter, the `Style` with the `geometry` configured as I showed above, and you'll need to use that style on both the Draw interaction and the Vector layer.

**Reply by Adeeljavedvirk** [reply]
Got it

**Reply by ahocevar** [reply]
When using the geometry function for the Draw interaction, you also need to handle the case where the geometry is a point, i.e. the cursor point:
```js
const ellipseStyle = new Style({
  geometry(feature) {
    const geometry = feature.getGeometry();
    if (geometry.getType() === "LineString") {
      const coordinates = geometry.getCoordinates();
      const center = coordinates[0];
      const dx = coordinates[1][0] - coordinates[0][0];
      const dy = coordinates[2][1] - coordinates[0][1];
      const circle = new Circle(center, dx);
      const polygon = fromCircle(circle, 64);
      polygon.scale(1, dy / dx);
      return polygon;
    } else {
      return geometry;
    }
  },
  stroke: new Stroke({
    color: "red",
    width: 2
  }),
  image: new CircleStyle({
    radius: 5,
    fill: new Fill({
      color: "red"
    })
  })
});

**Comment by ahocevar** [comment]
Complete working example: https://codesandbox.io/s/serverless-water-ntcvwx?file=/src/index.js

**Reply by Adeeljavedvirk** [reply]
Last help .. I want to draw different shapes, If i add the rectangle shape after the ellipse then ellipse shape disappear from the vector. Can you please update the working example with two shapes. Rectangle and ellipse.

**Reply by Adeeljavedvirk** [reply]
![image](https://github.com/openlayers/openlayers/assets/19381997/20a561c5-5be0-418d-899c-1cc2783dbe81)
I am getting the ellipse shape with every other shape

**Reply by ahocevar** [reply]
I think you should be able to figure out the rest form here. For rectangles, just refer to the https://openlayers.org/en/latest/examples/draw-shapes.html example. Instead of the `geometryFunction` above, use the built-in `createBox` function. The resulting geometry will be a polygon, so the style will continue to work, because it only modifies LineString geometries.
